### PR TITLE
E2E settings

### LINKED ISSUE #66649: refactor test/e2e/framework
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

**What happened**:

The E2E framework has grown organically over time. As a result, it is both hard to maintain and close to impossible to use outside of Kubernetes itself.

**What you expected to happen**:

The goal is to make it possible that other projects can:
- import test/e2e/framework as a vendor dependency without pulling in support for cloud providers and any tests (because of the huge dependency chains that a complete import causes)
- import support for cloud providers and tests separately if desired
- add custom tests without modifying the framework
- create and run a Ginkgo suite

One potential user for this is the Kubernetes CSI project, which currently relies on tests in Kubernetes to test the out-of-tree sidecar containers. Testing the "canary" sidecar containers is done manually by developers. There is an on-going effort to unify all sidecar containers in one repo (https://docs.google.com/document/d/1xC0Cm11WTebT9m_9yZbfzud24GJRxlp55gdR5QaMHmM/edit), which then could also import and run the E2E testing in TravisCI.

Other users are CSI driver developers.

For a full description of this proposal including an implementation plan, see this Google doc: https://docs.google.com/document/d/1lPG264ZlwPbIuFeP9P0-c0vXdmCC7wzwaapf53BbXBA/edit?usp=sharing

### PR DESCRIPTION
**Pull request by pohly**
**What this PR does / why we need it**:

As described in issue #66649, the goal is to make `test/e2e/framework` usable for test suites and tests outside of Kubernetes. External tests cannot modify the upstream framework when adding new configuration options, because `test/e2e/framework/test_context.go` is read-only for them.

There's also the stalled effort to support a configuration file format. Viper was chosen for that years ago, but then existing tests were never modified to retrieve options from Viper. Note that Viper might not be the right tool for the job. Other components (kubeadm, see https://github.com/kubernetes/kubernetes/issues/66649#issuecomment-418523536) also need something and the trend seems to go towards a Kubernetes-specific implementation.

This PR avoids making tests depend on Viper, it just uses Viper under the hood. This means that it could get replaced later without changing tests.

The approach suggested in this PR is (see comment in test/e2e/test_context.go):
* tests define their options using the `flag` package, in their own code, or use a new utility package to define flags via struct tags (requires less code)
* all flags can also be set in a configuration file if that feature gets enabled in a test suite
* test/e2e.test supports a config file, now for all options

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Partial #66649

**Special notes for your reviewer**:

This is a subset of PR #68483.

**Release note**:

```release-note
test/e2e/e2e.test:
* -viper-config can be used to set also the options defined by command line flags
* the default config file is "e2e.yaml/toml/json/..." and the test starts when no such config is found (as before) but if -viper-config is used, the config file must exist
* -viper-config can be used to select a file with full path, with or without file suffix
* the csiImageVersion/Registry flags were renamed to storage.csi.imageVersion/Registry
```

/sig testing

### CODE REVIEW

**Comment by pohly** [comment]
Despite the "WIP" tag the changes are already complete enough to be tested.

File: test/e2e/framework/viperconfig/viperconfig.go
Comment (neolit123) [inline test/e2e/framework/viperconfig/viperconfig.go]: minor nit; can be:
```
if !viper.IsSet(f.Name) {
	return
}
```
to avoid indentation of the whole block bellow.

File: test/e2e/framework/viperconfig/viperconfig.go
Comment (neolit123) [inline test/e2e/framework/viperconfig/viperconfig.go]: possibly combine multiple variables in a single line:
```
fmt.Sprintf("%v", viper.Get(f.Name))
```

File: test/e2e/framework/config/config.go
Comment (neolit123) [inline test/e2e/framework/config/config.go]: i like the approach.
 it creates some boilerplate, but overall this seems to comply with the demands of individual tests?

i will defer to the maintainers though for more reviews.

**Review by neolit123 — APPROVED** [review]
thanks @pohly for working on this.

i'm must say that the overall approach SGTM.
so this gets my +1, based on a quick review.

**Review by coffeepac — APPROVED** [review]

**Review by timothysc — CHANGES_REQUESTED** [review]

File: test/e2e/framework/viperconfig/viperconfig.go
Comment (pohly) [reply test/e2e/framework/viperconfig/viperconfig.go]: Done.

File: test/e2e/framework/viperconfig/viperconfig.go
Comment (pohly) [reply test/e2e/framework/viperconfig/viperconfig.go]: Done.

File: test/e2e/framework/config/config.go
Comment (pohly) [reply test/e2e/framework/config/config.go]: With "blurb about this state of versioning support", do you mean adding a warning that renaming fields or the common prefix is a user-visible change? I've added that and also updated the example a bit:
* the example, including the corresponding command line flags, is presented completely before moving towards the definition of the underlying mechanism
* the "scaling" part has to be defined in the prefix, because the struct name cannot be determined automatically (in this example, the struct is anonymous), nor is that desirable (it could be something that has no meaning outside of the source code)

File: test/e2e/framework/config/config.go
Comment (pohly) [reply test/e2e/framework/config/config.go]: No, not here. It's really https://godoc.org/flag and not https://godoc.org/github.com/spf13/pflag. `test/e2e` parses the command line with the former, in contrast to `test/e2e_node`, which uses the latter. So the "lingua franca" for defining flags that work everywhere has to be the `flag` package from the Go library.

File: test/e2e/framework/config/config.go
Comment (pohly) [reply test/e2e/framework/config/config.go]: Done.

**Comment by pohly** [comment]
I've updated the Viper config support:
- the default is as it was before this PR: look for "e2e", proceed without it
- the `-viper-config` command line flag is defined by the test suite, not by the package: cleaner, allows reuse
- the package returns errors instead of panicking (here I agree with @timothysc that returning an error is better because this error is probably caused by the user and thus may have to be reported differently)
- no more second parsing of command line flags with pflag, instead the priority of "flags over config" gets ensured by `viperUnmarshal` itself: this is simpler and solves a bug where `-provider foo` triggered an error from `pflag` (it treats single-dash options differently than `flag`)
- better error messages, for example: `viper config "e2e" = "/nvme/gopath/src/k8s.io/kubernetes/e2e.yaml": setting v from config file value: strconv.Atoi: parsing "foobar": invalid syntax` when setting "v", the integer log level, to a string in the e2e.yaml file

I think I addressed the review comments, but please check.

File: test/e2e/framework/config/config.go
Comment (timothysc) [reply test/e2e/framework/config/config.go]: I'm ok with it for now, but not a huge fan as it can be an anti-pattern is downstream folks import to leverage.

**Review by timothysc — APPROVED** [review]
/lgtm 
/approve

**Comment by deads2k** [comment]
I came across this today and I was surprised that UpgradeTarget and UpgradeImage are not considered core to the framework that runs all tests.  If different tests have different notions of the upgrade target and image, they aren't likely to work well together.

@smarterclayton

File: test/e2e/e2e_test.go
Comment (deads2k) [inline test/e2e/e2e_test.go]: This sort of distributed registration makes it really hard to depend on packages because of conflicts.  It also makes it impossible to avoid stripping vendor.  And it prevents visibility to any other thing that wants access.  I don't think we should use a pattern like this and instead we could use the `kubectl` or `kube-apiserver/controller-manager` method of `Options` structs with `Bind` methods.

Comment (pohly) [reply test/e2e/e2e_test.go]: The idea was to let each test own a certain parameter namespace (like 'storage' for test/e2e/storage). One of my earlier revisions of this PR described a naming scheme that would have avoided conflicts, but during review it was pointed out that this is too complex and therefore its now only a recommendation.

`UpgradeTarget` and `UpgradeImage` are legacy options that for historic reasons are defined at the top level of the namespace and weren't renamed to avoid breaking jobs. They don't belong into the core framework because they only parametrize certain tests.

Comment (deads2k) [reply test/e2e/e2e_test.go]: This has actually made it harder to build tests on top.  Instead of having a publicly exposed config struct that I can set any way I want and then use, I must have my executable called to set these registered flags.

Comment (neolit123) [reply test/e2e/e2e_test.go]: i see viper as a temporary solution here.
the work that WG component standard are doing is planned to solve the flags vs config problem, hopefully.
https://github.com/kubernetes/community/tree/master/wg-component-standard

one of the early steps in flight is a KEP for a fork of pflag (kflag):
https://github.com/kubernetes/enhancements/pull/764

Comment (pohly) [reply test/e2e/e2e_test.go]: > This has actually made it harder to build tests on top. Instead of having a publicly exposed config struct that I can set any way I want and then use,

I don't understand the part about "harder to build tests". Previously it was impossible to build tests on top of the framework without modifying the framework. Now that part works.

> I must have my executable called to set these registered flags.

Yes, of course the executable must be called, because it is the one parsing flags and/or configuration. Can you explain how and where you want to set the flags without involving the test suite binary?

I'm not trying to defend Viper here.  I've never liked it myself (no error messages when spelling an option wrong!) and if there is a better solution, then I'm sure that would be welcome.

Comment (deads2k) [reply test/e2e/e2e_test.go]: For example, I could previously choose take the configuration any way I wanted and set a value like this: https://github.com/openshift/origin/blob/master/cmd/openshift-tests/openshift-tests.go#L251 and we cannot anymore.

Comment (deads2k) [reply test/e2e/e2e_test.go]: You can see a similar example for the upgrade targets here: https://github.com/openshift/origin/blob/master/cmd/openshift-tests/upgrade.go#L57-L58 .  This pull has remove the logical separation of configuration from how that configuration is specified in a way that appears to be inconsistent with the direction of the rest of the project.

See patterns like this https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/options/authentication.go#L39 and https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cli-runtime/pkg/genericclioptions/builder_flags.go#L106 which allow separation, flexibility, and composition.

Comment (pohly) [reply test/e2e/e2e_test.go]: > set a value like this: https://github.com/openshift/origin/blob/master/cmd/openshift-tests/openshift-tests.go#L251

That sets the viper configuration file and then expects some code to do something with it. In 1.12, that code looked like this: https://github.com/kubernetes/kubernetes/blob/609a4b9323a752fc0e9a51d310cc836baa037a20/test/e2e/framework/test_context.go#L325-L332

In master, Viper support is no longer part of the framework itself. That's exactly the flexibility that you are talking about: instead of mandating one particular config parser, let the caller decide. To get a similar effect, you now need to do something like this: https://github.com/kubernetes/kubernetes/blob/ad4727440cc4a5be5c9714ffc0baefc11ce6c7f6/test/e2e/e2e_test.go#L59-L67

So the API has changed, but I think you can still do what you want to do.

> https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/options/authentication.go#L39

That's exactly how I would also do it. I even wrote utility code that automatically registers struct fields as individual parameters:
https://github.com/kubernetes/kubernetes/blob/ad4727440cc4a5be5c9714ffc0baefc11ce6c7f6/test/e2e/framework/config/config.go#L17-L34

Comment (deads2k) [reply test/e2e/e2e_test.go]: You can see this doesn't compose, right?  I cannot import the `test/e2e/lifecycle` package without it claiming global flags.  Then, because the flag is treated as an internal global variable, it's impossible to set externally and embed the included tests.

This is substantively worse than the composition described above where it possible to create the options, then set  them based on a config of the composer's choosing, then execute.

Comment (deads2k) [reply test/e2e/e2e_test.go]: > I don't think this leads to clean composition. I cannot import the `test/e2e/lifecycle` package without it claiming global flags. Then, because the flag is treated as an internal global variable, it's impossible to set externally and embed the included tests.

turns out you're hitting the same problem with klog: https://github.com/kubernetes/kubernetes/issues/75403 .  This pull created the same problem with the tests.  They're not importable anymore

Comment (pohly) [reply test/e2e/e2e_test.go]: >  I cannot import the test/e2e/lifecycle package without it claiming global flags.

You also can't import the `e2e/framework` either without it claiming global flags. That's actually worse, because one always gets the flags even when not using the corresponding tests. My proposal ("proposal" because not completely implemented) makes that better by only defining flags when actually importing the tests that use them.

Anyway, I thought that using `flag` and its default `FlagSet` was an acceptable mode of operation. That was fine for the Kubernetes E2E suite and no-one mentioned other usages where that might not be appropriate.

It's not hard to change. Instead of allowing tests to add to the default `FlagSet`, we define a `framework/config.FlagSet` that is to be used for test config options. Then the user of the framework can decide whether it wants to expose those as actual command line options or only wants to set them indirectly through some other mechanism, like `viperconfig`.

Comment (deads2k) [reply test/e2e/e2e_test.go]: > You also can't import the `e2e/framework` either without it claiming global flags. That's actually worse, because one always gets the flags even when not using the corresponding tests. My proposal ("proposal" because not completely implemented) makes that better by only defining flags when actually importing the tests that use them.

Before I had to explicitly call a method to register the flags (`RegisterCommonFlags` as I recall).  Now you're doing it on import.  Before I could import a package and choose how to use it, now I can't import the package.

Comment (pohly) [reply test/e2e/e2e_test.go]: > Before I had to explicitly call a method to register the flags (RegisterCommonFlags as I recall). Now you're doing it on import.

Right, I had forgotten that aspect. So you are asking for tests to register their flags only when called explicitly, right?

What about the Ginkgo tests? They also get registered during init, unconditionally. It's perhaps less likely that one wants to import a test package without actually using the tests, but it could happen. Someone might want to have finer control over which tests get registered, based on command line flags. We actually do have this issue right now for CSI testing: we use the Kubernetes E2E suite, but run only 28 out of 6740 tests. Loading all those JUnit files with the Skip entries is noticable slow in Spyglass.

Let me see whether I can come up with a solution...

**Comment by justinsb** [comment]
This is now the sole consumer of viper, after https://github.com/kubernetes/kubernetes/pull/76780 viper will be the sole consumer of afero; viper is already the sole consumer of hashcorp/hcl (at least).

How is this actually used?  Do we need anything more than yaml & json support?  My experience has been that it's easier to just write a parser, rather than using viper...

**Comment by neolit123** [comment]
i think that viper support in the e2e framework was a mistake.
though removing it now will probably break consumers.

something that i proposed to @timothysc is that we simply start building a separate V2 binary e.g. `e2e-suite2.test` and re-do it's flags. the current flags for the suite are unmanageable. similar was done in the case of kubetest -> kubetest2.

as a side note, this type of piping of config fields to CLI flags was briefly discussed during this week's WG component standard meeting with @mtaufen and @stealthybox .

**Comment by BenTheElder** [comment]
if we redo the e2e suite I'd love to see a _lot_ more redone than the flags
:^)

**Pull request merged**
